### PR TITLE
Automate release - WIP

### PR DESCRIPTION
**Pull request by bhamail**
Work on automating releases via CircleCI
The release stuff is currently only going to happen when a human triggers a release (and only on the 'master' branch too), so this PR shouldn't cause fireworks right away.

Also some GitHub template fixes (see the narly header above for why - shows in raw comment text). :)

Still needs more work to validate, and will likely extract stuff into an external CircleCI ORB, but this gets things started.

cc @bhamail / @DarthHater / @guillermo-varela

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -165,11 +174,16 @@ if (project.hasProperty('deployment')) {
       }
 
       signing {
+        useGpgCmd()
+        allprojects { ext."signing.gnupg.executable" =  System.getenv('GPG_EXECUTABLE') }
+        allprojects { ext."signing.gnupg.keyName" = System.getenv('GPG_KEYNAME') }
+        allprojects { ext."signing.gnupg.passphrase" = System.getenv('GPG_PASSPHRASE') }
```
Comment (guillermo-varela) [inline build.gradle]: Using this the same as:
```
allprojects {
  ext."signing.gnupg.executable" =  System.getenv('GPG_EXECUTABLE')
  ext."signing.gnupg.keyName" = System.getenv('GPG_KEYNAME')
  ext."signing.gnupg.passphrase" = System.getenv('GPG_PASSPHRASE')
}
```

But with just one iteration?

File: build.gradle [hunk 2]
```diff
@@ -109,8 +118,8 @@ if (project.hasProperty('deployment')) {
         def snapshotsRepoUrl = deployment.snapshot.url
         url = version.endsWith('SNAPSHOT') ? snapshotsRepoUrl : releasesRepoUrl
         credentials {
-          username = version.endsWith('SNAPSHOT') ? deployment.snapshot.username : deployment.release.username
-          password = version.endsWith('SNAPSHOT') ? deployment.snapshot.password : deployment.release.password
+          username = System.getenv('SONATYPE_USERNAME')
+          password = System.getenv('SONATYPE_PASSWORD')
```
Comment (guillermo-varela) [inline build.gradle]: Is it possible for any person to create a PR, print this values and see them on the CircleCI UI?

File: build.gradle (hunk 1 above)
Comment (bhamail) [reply build.gradle]: Yes!  Thank you.  e570439

File: build.gradle (hunk 2 above)
Comment (bhamail) [reply build.gradle]: Great question. A few things are in play that should keep these values safe.

1. The variables are only provided via a CircleCI [context](https://circleci.com/docs/2.0/contexts/#overview). The 'context' in use is [restricted](https://circleci.com/docs/2.0/contexts/#restricting-a-context). Basically, not everyone will be able to run a build that uses this 'context'.

1. CI is configured to only run 'release' builds (which require access to the restricted 'context') on the 'master' branch.

1. As a secondary safeguards, CircleCI will do [secrets masking](https://circleci.com/docs/2.0/contexts/#secrets-masking).

So...I think it is safe, but if not, we'll fix it.

**Pull request merged**